### ISSUE TITLE
PeekableReadBuffer: Memory limit exceed when inserting data by HTTP (20.8)

### ISSUE THREAD

**Issue by astudnev** [ClickHouse/ClickHouse#18690]
Error started to appear after migrated 19.9 -> 20.8

Settings
<input_format_parallel_parsing>0</input_format_parallel_parsing>
and
<input_format_values_interpret_expressions>0</input_format_values_interpret_expressions>

do not help

```
2021.01.03 08:01:15.601198 [ 1968 ] {b4472a5e-e463-4852-a0ed-efd90fae6393} <Error> DynamicQueryHandler: Code: 241, e.displayText() = DB::Exception: PeekableReadBuffer: Memory limit exceed, Stack trace (when copying this message, always include the lines below):

0. Poco::Exception::Exception(std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> > const&, int) @ 0x18cd2050 in /usr/bin/clickhouse
1. DB::Exception::Exception(std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> > const&, int) @ 0xe63232d in /usr/bin/clickhouse
2. ? @ 0x164be2e2 in /usr/bin/clickhouse
3. DB::PeekableReadBuffer::peekNext() @ 0x164bd633 in /usr/bin/clickhouse
4. DB::PeekableReadBuffer::nextImpl() @ 0x164bd978 in /usr/bin/clickhouse
5. void DB::readQuotedStringInto<true, DB::PODArray<char8_t, 4096ul, Allocator<false, false>, 15ul, 16ul> >(DB::PODArray<char8_t, 4096ul, Allocator<false, false>, 15ul, 16ul>&, DB::ReadBuffer&) @ 0xe67e4f5 in /usr/bin/clickhouse
6. DB::DataTypeString::deserializeTextQuoted(DB::IColumn&, DB::ReadBuffer&, DB::FormatSettings const&) const @ 0x156445a4 in /usr/bin/clickhouse
7. DB::ValuesBlockInputFormat::readRow(std::__1::vector<COW<DB::IColumn>::mutable_ptr<DB::IColumn>, std::__1::allocator<COW<DB::IColumn>::mutable_ptr<DB::IColumn> > >&, unsigned long) @ 0x16597b4b in /usr/bin/clickhouse
8. DB::ValuesBlockInputFormat::generate() @ 0x165982cd in /usr/bin/clickhouse
9. DB::ISource::work() @ 0x163e821b in /usr/bin/clickhouse
10. DB::InputStreamFromInputFormat::readImpl() @ 0x163ab46d in /usr/bin/clickhouse
11. DB::IBlockInputStream::read() @ 0x155611cd in /usr/bin/clickhouse
12. DB::InputStreamFromASTInsertQuery::readImpl() @ 0x15973f79 in /usr/bin/clickhouse
13. DB::IBlockInputStream::read() @ 0x155611cd in /usr/bin/clickhouse
14. DB::copyData(DB::IBlockInputStream&, DB::IBlockOutputStream&, std::__1::atomic<bool>*) @ 0x1558415e in /usr/bin/clickhouse
15. DB::executeQuery(DB::ReadBuffer&, DB::WriteBuffer&, bool, DB::Context&, std::__1::function<void (std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> > const&, std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> > const&, std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> > const&, std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> > const&)>) @ 0x15d06285 in /usr/bin/clickhouse
16. DB::HTTPHandler::processQuery(DB::Context&, Poco::Net::HTTPServerRequest&, HTMLForm&, Poco::Net::HTTPServerResponse&, DB::HTTPHandler::Output&) @ 0x16344c69 in /usr/bin/clickhouse
17. DB::HTTPHandler::handleRequest(Poco::Net::HTTPServerRequest&, Poco::Net::HTTPServerResponse&) @ 0x1634871b in /usr/bin/clickhouse
18. Poco::Net::HTTPServerConnection::run() @ 0x18bb1df3 in /usr/bin/clickhouse
19. Poco::Net::TCPServerConnection::start() @ 0x18bf000b in /usr/bin/clickhouse
20. Poco::Net::TCPServerDispatcher::run() @ 0x18bf0728 in /usr/bin/clickhouse
21. Poco::PooledThread::run() @ 0x18d6ee36 in /usr/bin/clickhouse
22. Poco::ThreadImpl::runnableEntry(void*) @ 0x18d6a230 in /usr/bin/clickhouse
23. start_thread @ 0x7494 in /lib/x86_64-linux-gnu/libpthread-2.24.so
24. clone @ 0xe8aff in /lib/x86_64-linux-gnu/libc-2.24.so
 (version 20.8.11.17 (official build))

```

**Comment by den-crane** [comment]
Do you have reproducible example? DDL ? Data sample? Steps?

**Comment by astudnev** [comment]
```
CREATE TABLE IF NOT EXISTS bitcoinsv_etl.outputs ON CLUSTER replicated3
(
  tx_date Date,
  tx_time DateTime,

  blockchain_id UInt32,
  block UInt32,

  tx_id String,
  tx_index UInt16, -- 0-based index of transaction in block

  output_index UInt16, -- 0-based index of output in transaction

  address String,  -- Base58 output address

  script_type LowCardinality(String),
  script_template UInt32, -- from dictionary script_templates
  script String,
  req_sigs UInt32, -- ReqSigs field

  value Decimal64(8), -- output amount in BTC

  direction Enum8('unknown' = 0, 'not_change' = 1, 'change' = 2, 'likely_not_change' = 3, 'likely_change' = 4, 'mining' = 5, 'fee' = 6)

) ENGINE = ReplicatedMergeTree('/clickhouse/archive_servers/tables/bitcoinsv_etl/outputs', '{replica}')
  ORDER BY block
  TTL tx_date + INTERVAL 1 MONTH
  SETTINGS index_granularity=8192;
```


```
INSERT INTO bitcoinsv_etl.outputs(tx_date,tx_time,blockchain_id,block,tx_id,tx_index,output_index,address,script_type,script_template,script,req_sigs,value,direction) VALUES
```

[data_1609667508.txt](https://github.com/ClickHouse/ClickHouse/files/5762399/data_1609667508.txt)

**Comment by den-crane** [comment]
curl 'http://localhost:8123/?query=INSERT+INTO+outputs(tx_date,tx_time,blockchain_id,block,tx_id,tx_index,output_index,address,script_type,script_template,script,req_sigs,value,direction)+VALUES ' --data-binary @data_1609667508.txt
Code: 241, e.displayText() = DB::Exception: PeekableReadBuffer: Memory limit exceed (version 20.13.1.5579)




echo 'INSERT INTO outputs(tx_date,tx_time,blockchain_id,block,tx_id,tx_index,output_index,address,script_type,script_template,script,req_sigs,value,direction) VALUES'`cat data_1609667508.txt ` |clickhouse-client

SELECT count() FROM outputs
┌─count()─┐
│    9330 │
└─────────┘

**Comment by tavplubix** [comment]
The error happens because data contain huge string field (at row 8244), it's about 18 MiB. `PeekableReadBuffer` has limitation of maximum 16 MiB for one field, because in most cases fields of such size mean that something is wrong with parsing (for example, quotes are not closed, so parser reads everything until EOF as one field). We can increase or remove this limitation. However, it's usually a bad practice to store blobs in database.

**Comment by astudnev** [comment]
May be better to configure it?

**Comment by alexey-milovidov** [comment]
@tavplubix Let's increase it to 1 GiB or remove completely (rely on memory limit).

> However, it's usually a bad practice to store blobs in database.

I challenge this limitation in this PR #18842 - there are BLOBs of 500 MiB and everything is ok so far...

**Comment by tavplubix** [comment]
Ok.
>> However, it's usually a bad practice to store blobs in database.
>
> there are BLOBs of 500 MiB and everything is ok so far...

We had an issue with blobs, it might cause merges to fail with memory limit on huge granules. It's almost fixed in #17120, but AFAIK it's still reproducible in very rare cases.
